Add which-key system

Cretezy opened the pull request:
Closes https://github.com/zed-industries/zed/issues/10910

This is a work in progress implementation, but it functionally works. I'm looking for some feedback on both implementation and design.

This PR adds a `which-key`-like system, which shows possible bindings to take after starting a sequence of key which leads to a binding. This is mostly useful in Vim mode, as there's multiple common sequences (such as `g` which leads to `gv`, `gc`, ..., or `[`/`]`, etc). Currently, this shows in the main screen area (doesn't draw over panels), to be less intrusive.

Common use cases are to see what can follow `ctrl+k`, `g`, `z`, `ctrl+w`, or `space`/leader. `g`/`ctrl+w` are worst case scenarios, as they have tons of features built-in.

<img width="3065" height="1633" alt="image" src="https://github.com/user-attachments/assets/4e5135cf-495d-456f-94b8-856b9aa90284" />



`"location": "left_panel"`:
<img width="3063" height="1638" alt="image" src="https://github.com/user-attachments/assets/bd190188-f89b-4b08-a10c-51c9ba4ea0c0" />



To do:
- [ ] Add ability to scroll menu with keyboard


Additionally, there's some follow-up work which could enable this menu when holding down `ctrl` for example. This would make this feature much more useful in non-Vim mode.

Release Notes:

- Add which-key system to show possible next actions for sequences of bindings (e.g. `ctrl-w`)

Cretezy committed:
Add which-key system

Cretezy commented on crates/command_palette/src/command_palette.rs:
```diff
@@ -437,30 +437,6 @@ impl PickerDelegate for CommandPaletteDelegate {
     }
 }
 
-pub fn humanize_action_name(name: &str) -> String {
```
Had to move this to avoid circular dependency, not sure if `gpui` is the right destination.

Cretezy commented on crates/gpui/src/window.rs:
```diff
@@ -3620,28 +3620,30 @@ impl Window {
-                    window.replay_pending_input(to_replay, cx)
-                })
-                .log_err();
+                // TODO: This resets the current pending inputs after 1s. This doesn't really work with a which-key system.
```
Looking for feedback for what to do with this. which-key is not usable if the pending keys are cleared after 1s.

Cretezy replied:
As per discuss below, I will implement that this is disable if the which-key menu is open.

Cretezy commented:
@theherk Good call for showing only the next vs showing all. I'll add support for that and make it configurable.

I've already got font weight for keys locally, plus some other minor design/usability chances + configuration options. I'll push that tomorrow and update the screenshots.

rodrigopex commented:
Hey @Cretezy, this is a wanted feature. Congrats on submitting that! In terms of UI, I recommend increasing highlight in the key and reducing it in the description. I suppose you could add some text or indication to make the current key clearer.

I'm not sure how difficult it would be to group things. In your case, we can see when you press `ctrl-w`, there are `g space`, `g t`,  `g shift-d`. These may appear in a single one under `g`, and when pressing `g`, the rest of the options could appear. At least you could make the list in alphabetical order to create a sequence. There are the actions and groups of actions under `g`, for example. 

Take a look at the Lazyvim version of that. 
<img width="361" height="656" alt="image" src="https://github.com/user-attachments/assets/e781ee6c-8072-4f41-bae2-7581cf2b08c7" />

I hope you can finish that and your PR is merged soon. Thanks a lot.

Cretezy committed:
Add configuration, update design

Cretezy commented:
@domi413 Yes those improvements are planned, still working on the design!

ConradIrwin commented:
Nice! CC @danilo-leal for some design love.

I think we only really need one setting (which is disabled by default), and which shows the data in what we think is the best way (not yet sure if that's the full key sequence or just the next key). One thing that's maybe different in Zed vs Helix is that we have *way* more shortcuts showing up in here, so if we can figure out how to keep the UI from taking over the entire screen it would be nice...

In terms of the 1s timeout, we get a lot of feedback about that. We could try removing it completely, but I'm worried about people hitting `ctrl-w` and then switching task and hitting a key shortcut that doesn't do what they want.. We could also make it much longer, not sure if there are other examples of how apps with lots of key sequences handle the issue.

Cretezy committed:
Group keybinds

ConradIrwin commented:
Hmm. Looking into vim, they have the 1s timeout when there are ambiguous key matches (e.g. do `:nmap g D`, then `g` after 1s it'll delete the line, or `:imap jk <esc>` after 1s it'll insert a j); but they don't have any for when you've typed a prefix and there is no conflict.

VScode seems closer to 5s by default, but I don't think you can have the key prefix matches (it seems to always use the shorter match instantly; though I may have missed how to configure the conflict correctly).

It seems reasonable to split the two cases up:

* In the common case that the prefix has no binding, default to a 5s timeout (to match vscode) or none (to match vim). If the which key menu is open the default should be none, because it makes it obvious in the UI that this is the state you're in; so maybe none is the right default here.
* In the uncommon case that the prefix does have a binding (or like `jk` in insert mode, the key will insert a `j` after a bit), keep the current 1s to match vim. Not sure if we should show the which key menu or not in that case.

Cretezy committed:
Use column layout, align keystrokes

Cretezy commented:
I've pushed another update which overhauls the design and introduces grouping. Still work to do, but I'm happy to collect feedback at this point.

Cretezy commented on crates/workspace/src/which_key_layer.rs:
```diff
@@ -0,0 +1,324 @@
+    h_flex().items_center().gap_1().children([
+        keystroke.into_any_element(),
+        div()
+            .child(Label::new("🡒").color(Color::Muted))
```
Likely best to replace this with a proper icon

Cretezy commented:
@bjorn-ove Agreed! I'm not sure if this belongs in the settings file or keymap file yet. I think keymap makes more sense, although it doesn't fit in the current model of the file.

If it's in the settings, it could be under `which_key.groups`, which could be a mapping of prefixes to group names. Additionally, we'd probably want to ship with names for the groups that exists in the default key mappings.

rodrigopex commented:
@Cretezy, another suggestion. Instead of `shift-d`, you could show `D`. The uppercase letters can be explicitly shown in uppercase, like which-key nvim:

<img width="343" height="183" alt="image" src="https://github.com/user-attachments/assets/cf1450e8-191c-4210-8626-eafd2738d700" />

Cretezy committed:
Add left_panel location

evolify commented:
Want this feature too.

Cretezy commented:
Hey! Although this PR isn't ready for production, I'd appreciate an early review from the Zed team on the implementation as well as some design feedback if some changes are required. I'm happy to communicate through GitHub comments, or I'm also available on Discord (same username).

ConradIrwin commented:
@Cretezy If you have some time to pair on this we can look at it together next week: https://cal.com/conradirwin/pairing.

I am wondering if this is actually the right way to solve the problem of "I don't know the keybinding". In Zed the best way to learn a shortcut for an action is to type it into the command palette, and you can see the binding:

<img width="601" height="175" alt="Screenshot 2025-08-13 at 22 40 20" src="https://github.com/user-attachments/assets/ead29556-ec95-4e89-a955-28de02948d83" />

This doesn't work in vim, so maybe which key is more useful there.

It seems a bit odd to me that someone would remember that "the keybinding is somewhere behind `g`". That said it may be self-reinforcing; if you always have whichkey, you never need to learn the second key of the shortcuts. Maybe that's a good thing, but it's different from how I think about things (which is to try and make shortcuts muscle memory).

I do know that lots of people want something like this, so if there are good reasons for it to exist that I'm missing, please do let me know.

Assuming we do want to ship this (though I'm pretty sure it would be default "off"), the main question in my mind is how to make it feel helpful and not overwhelming. I notice that LazyVim only includes a subset of the keys available, should we emulate that and use this as a way to help teach people the more useful shortcuts first? Or is there a way to more sensibly prioritize these lists.

<img width="428" height="402" alt="Screenshot 2025-08-13 at 22 38 05" src="https://github.com/user-attachments/assets/b1c7ea15-744b-413d-89d2-02660bbe7a04" />

We can also tidy up the visual design to reduce the padding and try and increase the information density which might help.

Cretezy committed:
Merge master

Cretezy commented:
Small update: Met with Connor (thank you!) and will be continue work on this in the following week. Will post another update once there's more to chat about, hoping to get this over the finish line soon.

asd-noor commented:
After a google search for such feature in Zed, I found this PR.
I am really enjoying using Zed but this is a feature that I am missing badly from my current Neovim setup. Any estimated timeline of this feature getting merged? It's okay if there isn't any.

Cretezy committed:
Merge master

Cretezy committed:
Adjust styling, add filtered keys

Cretezy committed:
Re-add timer

Cretezy committed:
Support zoom

Cretezy committed:
Add cursor-avoiding behavior

Cretezy committed:
Merge master

Cretezy committed:
Fix mapping of bindings

Cretezy committed:
Adjust styling

bgub commented:
Tried this out. Really enjoy it!

ConradIrwin commented on crates/workspace/src/which_key_layer.rs:
```diff
@@ -0,0 +1,531 @@
+use crate::Workspace;
```
@Cretezy If possible I'd like to remove any changes from the workspace crate (which is already a bit of a jumble).

Is it possible to make the WhichKeyLayer a modal? In that case we could show it only if there is no other modal showing, to avoid this:

<img width="1824" height="823" alt="Screenshot 2025-09-02 at 21 58 25" src="https://github.com/user-attachments/assets/986879af-5d71-4504-8f92-694f7910c505" />

The only thing that might cause problems is that the modal layer currently renders a bunch of space above the modal, but we can either add an item to the trait that disables that, or account for it in the positioning logic.

Cretezy replied:
Yes will take a stab at this

Cretezy replied:
How did you trigger that @ConradIrwin?

Cretezy committed:
Disable by default, re-add filtered keys

Cretezy committed:
Remove grouping

Cretezy committed:
Merge master

ConradIrwin commented:
To show a modal you can do:
```rust
if !workspace.has_active_modal() {
 workspace.toggle_modal(|window, cx| WhichKey::new(window, cx))
}
```

I'm still sure we shouldn't ship column mode. Two reasons:
1. It only works sometimes, so it jumps between bottom mode and side mode depending on the width of the column
2. It's extra work for anyone who's changing this code to test in two different configurations.

So yes, in the abstract, more flexibility is good; but it's better to have a tighter scoped feature that works well than a loose collection of things that kind of work.

Removing the timeout completely seems reasonable, it sounded like from your comments that that has effectively already happened, we just need to tidy the code up?

Cretezy committed:
Remove left_panel location

Cretezy committed:
Move to modal

Cretezy commented:
@ConradIrwin I've moved the implementation to a modal. It should basically behave as before (hides when there's no pending keys or when focus changes), plus it won't conflict with other modals. I've also removed the timer.

Cretezy commented on crates/workspace/src/workspace.rs:
```diff
@@ -1146,6 +1147,7 @@ pub struct Workspace {
     _items_serializer: Task<Result<()>>,
     session_id: Option<String>,
     scheduled_tasks: Vec<Task<()>>,
+    pub which_key: Option<AnyEntity>,
```
I'm not 100% sure if this is the correct approach to preventing `WhichKey` from being dropped.

ConradIrwin committed:
Fix modal dismiss / move logic out of render

ConradIrwin committed:
Allow which key to take over terminal

ConradIrwin committed:
Improve render performance

layout_as_root is quite expensive, so render() was taking ~16ms in a
debug build for `g`. Without too much loss of accuracy, we now just
layout as root once per column

ConradIrwin commented:
I pushed up a few changes, notably using a closure to retain the which key state so we really don't need to change anything in the workspace crate; and removing the new `hide_modal` method, and emitting the DismissEvent instead.

I also made some changes to render. It turns out that `layout_as_root` is quite expensive; and so I've updated the code do this once per column instead of once per item (taking the time to call render() on `g` from `16ms` to 500µs in a debug build).

Finally I realized that we *do* need to keep the default timer in the case of conflicting shorter bindings, for example if you have:

```
  {
    "context": "vim_mode == insert",
    "bindings": {
      "j k": ["workspace::SendKeystrokes", "escape"]
    }
  },
```

If you hit `j` in insert mode, and do nothing, it should *not* open which-key, but it should type the *j* after 1 second. I didn't fix this yet though.

Cretezy commented:
@ConradIrwin I think we can also revert the changes with `humanize_action_name` now as well, since it doesn't really belong in `gpui`

ConradIrwin commented:
Perfect. Are you going to take the next step, or should I?

Cretezy committed:
Merge master

Cretezy committed:
Revert humanize action name, fix padding issues, show next line

Cretezy commented:
@ConradIrwin I haven't, I'm still a bit stuck at the timeout. I booked some time on your calendar this Friday to pair, hopefully we can finish this up.

xipeng-jin commented on crates/which_key/src/which_key_settings.rs:
To reflect the new changes from the upstream:main branch. I made some changes in which key settings accordingly as follows:
_crates/which_key/src/which_key_settings.rs_
```rust
use serde::Deserialize;
use settings::{Settings, SettingsContent, SettingsKey, WhichKeySettingsContent};

#[derive(Deserialize)]
pub struct WhichKeySettings {
    #[serde(default)]
    pub enabled: bool,
    #[serde(default = "default_delay_ms")]
    pub delay_ms: u64,
}

fn default_delay_ms() -> u64 {
    700
}

impl Settings for WhichKeySettings {
    fn from_settings(content: &SettingsContent) -> Self {
        let which_key: WhichKeySettingsContent = content.which_key.clone().unwrap_or_default();

        Self {
            enabled: which_key.enabled.unwrap_or(false),
            delay_ms: which_key.delay_ms.unwrap_or_else(default_delay_ms),
        }
    }

    fn import_from_vscode(_vscode: &settings::VsCodeSettings, _current: &mut SettingsContent) {
        // No equivalent setting in VScode
    }
}

impl SettingsKey for WhichKeySettings {
    const KEY: Option<&'static str> = Some("which_key");
}
```


```rust
pub struct SettingsContent {
...
    /// Settings for the which-key popup.
    pub which_key: Option<WhichKeySettingsContent>,
    ...
}
...
/// Settings for configuring the which-key popup behaviour.
#[skip_serializing_none]
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize, JsonSchema, MergeFrom)]
pub struct WhichKeySettingsContent {
    /// Whether to show the which-key popup when holding down key combinations
    ///
    /// Default: false
    pub enabled: Option<bool>,
    /// Delay in milliseconds before showing the which-key popup.
    ///
    /// Default: 700
    pub delay_ms: Option<u64>,
}
```

xipeng-jin reviewed (changes requested):
Hello @Cretezy @ConradIrwin , thanks for all your contributions. I am very interested in making this PR to work so I came across all the commits and I suggest to make some changes here which I left them in the comments. The goal is to reflect the new changes in the upstream main branch in order to make the which-key feature working. Thank you.

eugenesvk commented:
User customization of the layout/order (for grouping similar commands together)/description would be a great improvement

For example, you could add icons to your commands in your user config "icon" field for better readability (so add an "icon" field to keybind config

(Similar to https://github.com/zed-industries/zed/pull/34798#issuecomment-3097673843)

Then you could also have a "cheatsheet" version popup first: instead of a huge table with a lot of text you could only show a "keyboard layout" view with icons, any only if that's still not enough you could mouseover for a detailed tooltip or press a key to show the full text table

the ability to search the table would also help in long tables

ConradIrwin commented:
I think we should probably pause on this branch (as it's going stale) and open a first PR to fix the key timeout problem.

Currently if you type a key that is a prefix of a keyboard shortcut and then don't finish the shortcut it is cancelled after 1 second. We should change that logic to keep it pending forever (or until escape is pressed) in the case that there is no binding (and no input). For things like `j` in insert mode, we do want to still insert that after 1s if you have bound `jk` to escape.

Once we have that problem solved, then the remaining work is "just render the UI" and we can re-use the code from this branch to do that.

Going to close for now, but I am still interested in PRs for this (as it is very highly requested), and I think the code here is heading the right direction (it's just very conflicted).
